### DISCUSSION TITLE
LineMaterialRealSampler with AD

### DISCUSSION THREAD

**Discussion by FloPasselaigue** [idaholab/moose#17857] [Q&A Tools] [answered]
Hello,

I am working on an update of the hydrogen in BISON, and so I developed a test case. For this test I check the value of the material 'solubility' along the x axis, using the VectorPostProcessor LineMaterialRealSampler:

```
[VectorPostprocessors]
  [solub]
    type = LineMaterialRealSampler
    property = 'solubility'
    sort_by = x
    start = '0 2e-4 0'
    end = '25e-3 2e-4 0'
    outputs = diff
  []
[]
[Outputs]
  [profile]
    type = CSV
    execute_on = 'final'
    create_final_symlink = true
  []
[]
```
The normal test works just fine, however the AD version fails, raising a `Bad file descriptor error`:

```
The following error occurred in the object "solubility", of type "LineMaterialRealSampler".
In LineMaterialSamplerBase material property: solubility does not exist.
```

The only difference between the two versions is that the argument `use_automatic_differentiation` is set to `true` (below is the part of interest in the test file)

```
  [solubility]
    type = 'CSVDiff'
    input = 'solubility.i'
    csvdiff = 'solubility_profile_solubility_FINAL.csv'
    rel_err = 1e-2
    requirement = "BISON shall be able to compute the effective solubility of hydrogen."
  []
  [solubilityAD]
    type = 'CSVDiff'
    input = 'solubility.i'
    csvdiff = 'solubility_profile_solubility_FINAL.csv'
    rel_err = 1e-2
    requirement = "BISON shall be able to compute the effective solubility of hydrogen with AD."
    prereq = solubility
    cli_args = 'CladdingHydrides/hydrides/use_automatic_differentiation=true'
  []
```
I tried with another material to output and ran into the same issue, so it doesn't seem to be specific to the solubility material. Maybe I missed something, but I don't think the material names are changed when using AD, or that the VectorPostprocessors classes need an AD equivalent? I have a similar test in which I check the profile of a variable using `LineValueSampler` and it works, so it seems to be specific to `LineMaterialRealSampler`.

If anyone has a solution I'd really appreciate it.
Thanks for your help,
Florian Passelaigue

**Comment by hugary1995** [comment, marked as answer]
`LineMaterialSamplerBase` is already templated. But `LineMaterialRealSampler` is only specialized for `Real` not `ADReal`. You would want to add a template parameter `bool is_ad` in `LineMaterialRealSampler`, and instantiate it to `LineMaterialRealSampler` when `is_ad = false`, and instantiate it to `ADLineMaterialRealSampler` when `is_ad = true`.

**Reply by FloPasselaigue** [reply]
Thank you for your answer, I took a look but I wouldn't know how to do this modification. I think I found another way to test the implementation without using `LineMaterialRealSampler`.

**Reply by GiudGiud** [reply]
You can look at the GenericConstantMaterial for examples of AD templating
